Add CCFileProcessorSparkJob to support file-wise processing

jt55401 opened the pull request:
For some spark jobs, we want to process an entire file at one time.
I copied and simplified sparkcc to do this.
This is used in the upcoming integrity process.

jt55401 committed:
Add sparkccfile.py to support file-wise processing in spark jobs (used in integrity job)

sebastian-nagel commented:
Thanks for the contribution, @jt55401!

First, I understand the use case: be able to process any kind of file, not only WARC files and derivatives (WAT, WET).
Yes, is is "more generic" and allows to cover more use cases. Including those where you want to
- process a WARC file without using a WARC parser
- use a custom output format.

Of course, this is already possible using CCSparkJob as base class by overriding the method run_job(...). For example,  [HostLinksToGraph](/commoncrawl/cc-pyspark/tree/main/hostlinks_to_graph.py) reads and writes from/to Parquet. From CCSparkJob it uses only command-line parsing / option processing, logging definitions and optional profiling.

> A Spark job definition to process individual files from a manifest.

Good catch! And very well expressed! This needs to be put into the README because it's one of the central design decisions. I never really thought about it, just ported it from [cc-mrjob](/commoncrawl/cc-mrjob). I do not know why @Smerity decided to use a manifest while most big data tools read the input list from command-line arguments. I see one advantage of the manifest: it's easy to select a (larger) random sample. Using command-line arguments you quickly may reach the system limit on the maximum argument length when passing, say, 1k paths to WARC files as arguments.

One question: What's the rationale for using a NamedTemporyFile? Being able to share the content as a file with other processes?

... and one remarks which should be addressed: 90% of the code lines in sparkccfile.py are copied unmodified from sparkcc.py. This complicates maintenance because contributors may forget to implement a bug fix or improvement in both files.

Two suggestions how to reduce the code duplication:

1. CCSparkJob inherits from CCFileProcessorSparkJob
   - (preferred variant, although more work; expected to remove more duplicated code) 
   - move the definition of CCFileProcessorSparkJob into sparkcc.py
     - makes the deployment easier
     - avoids that the deployment for existing setups is broken by the changes below
   - remove duplicated code from CCSparkJob
     - keep only code / methods specific for WARC file processing
   - (difficult) fetch_warc to call fetch_file where applicable
     - the method "fetch_warc" is complex (120+ lines of code)
     - "fetch_file" mostly duplicates 70 lines
     - ideally 

2. CCFileProcessorSparkJob inherits from CCSparkJob
   - (easier to implement)
   - cf. above and [HostLinksToGraph](/commoncrawl/cc-pyspark/tree/main/hostlinks_to_graph.py)
   - basically, only fetch_file(...) and run_job(...) are then implemented by CCFileProcessorSparkJob
   - could move the definition also into sparkcc.py for easier deployment

In any case: the job / class should be listed in the README, maybe together with a simple example.

jt55401 commented:
@sebastian-nagel  - thank you for the review, it's greatly appreciated.

> NamedTemporyFile

Yes - that is exactly right. Some parts of the jobs we run use external tools, so we need a file in the outside world.

> Manifests vs. command line args

Yes - we are commonly processing 250,000-1,000,000+ files in a run. (Example: all the wet/wat files for an entire year of crawls)

> Two suggestions how to reduce the code duplication:

Ah, very astute. I will review these options and update the PR with a refactor.

> In any case: the job / class should be listed in the README, maybe together with a simple example.

Yes, no problem.

jt55401 committed:
Merge CCFileProcessorSparkJob into sparkcc.py

jt55401 committed:
Add CCFileProcessorSparkJob example and link in readme for it.

jt55401 commented:
OK @sebastian-nagel - I'm reasonably happy with this version.

The only slight downside is due to the way it's packaged, we now have to depend on warcio when it's not really needed. I'm not a deep expert in python modularization, so, if there is a clever way to fix this while preserving the cleanliness of this refactor, please let me know - otherwise, I'm fine leaving it.

Let me know if you have any other feedback.

jt55401 committed:
fix s3 functions so they work in spark environment

jt55401 commented:
I've since enhanced this further with 3 more functions:
* validate_s3_bucket_from_uri
* check_for_output_file
* write_output_file

These are mostly convenience functions which do as they each say for local file paths or S3 paths.
This further makes writing jobs which process entire files, and output NEW files to locations other than the default spark result table (which I've been using more as an audit log for such file-wise processing jobs)

jt55401 committed:
fix bug when the file was not downloaded from s3

jt55401 committed:
fix: fix scheme to recognize s3a and s3n

jt55401 committed:
fix: don't log 404 errors when checking for existence

sebastian-nagel committed:
docs: update description of MD5Sum job

sebastian-nagel committed:
refactor: Python code style

sebastian-nagel committed:
fix(CCFileProcessor): catch exceptions while opening local files,
                      properly count successfully or failed processed files

sebastian-nagel committed:
refactor(CCFileProcessor): rename command-line option

Rename command-line option `--output_base_uri` to
`--output_base_url` following the equivalen option
`--input_base_url`.

Improve logging of counters / aggregators.

sebastian-nagel committed:
refactor(CCFileProcessor): remove unused methods and code

- remove unused method `validate_s3_bucket_from_uri`
- remove code checking for output location on http(s)
  because it's not implemented to write output over HTTP

sebastian-nagel committed:
Bundle functionality shared between methods of CCSparkJob
and CCFileProcessorSparkJob to reduce duplicated code

sebastian-nagel committed:
CCFileProcessor: Raise an exception if input or output
                 location uses an unsupported scheme

- fail for hdfs:// input and output, not implemented yet.
- fail for http:// or https:// output.

sebastian-nagel committed:
docs: improve descriptions of CCSparkJob and derived core classes

sebastian-nagel committed:
Add more FastWARC examples and complete README

malteos commented:
Not sure if I am qualified to review this but it generally looks good at a first glance. However, I did not run all the modified scripts myself. Can you share example commands that are expected to work?

At least, it seems to break the example from the README (or is this warning expected?):

```
$SPARK_HOME/bin/spark-submit \
    --packages org.apache.hadoop:hadoop-aws:3.3.2 \
    ./cc_index_word_count.py \
    --input_base_url s3://commoncrawl/ \
    --query "SELECT url, warc_filename, warc_record_offset, warc_record_length, content_charset FROM ccindex WHERE crawl = 'CC-MAIN-2020-24' AND subset = 'warc' AND url_host_tld = 'is' LIMIT 10" \
    s3a://commoncrawl/cc-index/table/cc-main/warc/ \
    myccindexwordcountoutput \
    --num_output_partitions 1 \
    --output_format json
```

```
5/09/22 13:45:58 WARN FileSystem: Failed to initialize filesystem s3a://commoncrawl/cc-index/table/cc-main/warc: java.lang.NumberFormatException: For input string: "60s"
25/09/22 13:45:58 WARN FileStreamSink: Assume no metadata directory. Error while looking for metadata directory in the path: s3a://commoncrawl/cc-index/table/cc-main/warc/.
java.lang.NumberFormatException: For input string: "60s"
        at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67)
        at java.base/java.lang.Long.parseLong(Long.java:709)
        at java.base/java.lang.Long.parseLong(Long.java:832)
        at org.apache.hadoop.conf.Configuration.getLong(Configuration.java:1607)
        at org.apache.hadoop.fs.s3a.S3AUtils.longOption(S3AUtils.java:1024)
        at org.apache.hadoop.fs.s3a.S3AFileSystem.initThreadPools(S3AFileSystem.java:719)
        at org.apache.hadoop.fs.s3a.S3AFileSystem.initialize(S3AFileSystem.java:498)
        at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:3615)
        at org.apache.hadoop.fs.FileSystem.access$300(FileSystem.java:172)
        at org.apache.hadoop.fs.FileSystem$Cache.getInternal(FileSystem.java:3716)
        at org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:3667)
        at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:557)
        at org.apache.hadoop.fs.Path.getFileSystem(Path.java:366)
        at org.apache.spark.sql.execution.streaming.FileStreamSink$.hasMetadata(FileStreamSink.scala:55)
        at org.apache.spark.sql.execution.datasources.DataSource.resolveRelation(DataSource.scala:381)
        at org.apache.spark.sql.catalyst.analysis.ResolveDataSource.org$apache$spark$sql$catalyst$analysis$ResolveDataSource$$loadV1BatchSource(ResolveDataSource.scala:143)
        at org.apache.spark.sql.catalyst.analysis.ResolveDataSource$$anonfun$apply$1.$anonfun$applyOrElse$2(ResolveDataSource.scala:61)
        at scala.Option.getOrElse(Option.scala:201)
```

What about adding some unit test to ensure all examples work (especially on all the different platforms)?

malteos commented:
@sebastian-nagel I managed to run the example with `pyspark==3.5.6`. 

Should we add this to the `requirementst.txt` to make the version requirements more explicit?

```
pyspark>=3.2.3,<=3.5.6
```

To run the command locally, I also needed to add this Spark config to the command:

```
 --conf spark.driver.host=localhost  --conf spark.driver.bindAddress=localhost
```

Also, I needed to explicitly export the AWS secrets as ENV vars. Maybe we should add something like this to the README (even though it's already explained in the Hadoop docs):

```bash
  export AWS_ACCESS_KEY_ID=$(aws configure get aws_access_key_id)
  export AWS_SECRET_ACCESS_KEY=$(aws configure get aws_secret_access_key)
  export AWS_SESSION_TOKEN=$(aws configure get aws_session_token)  # if using temporary credentials
```

sebastian-nagel committed:
Refactor WET extractor to use identified charset
and language from the metadata record

Transfer information from warcinfo record

sebastian-nagel committed:
Experimental WARC-to-WET extractor

Add example tool relying on CCFileProcessorSparkJob extracting
text content from WARC HTML records and storing the texts in WET files.
For every WARC file one WET file is created. The WET record locations
are written to the output table.

sebastian-nagel committed:
fix(README): update Hadoop version and instructions about S3/S3A support

Add schema merging option to example command-line call.

sebastian-nagel committed:
Add shell script to launch cc-pyspark jobs on Hadoop
(Spark on YARN).

sebastian-nagel reviewed (approved):
Everything looks good
- all tools tested locally
- WET extraction (based on CCFileProcessorSparkJob) tested also on cluster

I'm going to merge the PR. Thanks, @jt55401 and @malteos!
